integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/predicate/NotPredicateSpecificsIT.java

```diff
@@ -0,0 +1,179 @@
+	private static final Integer FIELD5_VALUE3 = 142_000; // Different from document 1
+
+	@ClassRule
+	public static final SearchSetupHelper setupHelper = new SearchSetupHelper();
+
+	private static final SimpleMappedIndex<IndexBinding> index = SimpleMappedIndex.of( IndexBinding::new );
+
+	@BeforeClass
+	public static void setup() {
+		setupHelper.start().withIndex( index ).setup();
+
+		initData();
+	}
+
+	@Test
+	public void notMatchAll() {
+		assertThatQuery( index.query()
+				.where( f -> f.not( f.matchAll() ) ) )
+				.hasNoHits();
+	}
+
+	@Test
+	public void notMatchNone() {
+		assertThatQuery( index.query()
+				.where( f -> f.not( f.matchNone() ) ) )
+				.hasTotalHitCount( 3 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_1, DOCUMENT_2, DOCUMENT_3 );
+	}
+
+	@Test
+	public void notMatchSpecificValue() {
+		assertThatQuery( index.query()
+				.where( f -> f.not( f.match().field( "field2" ).matching( FIELD2_VALUE1 ) ) ) )
+				.hasTotalHitCount( 2 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_2, DOCUMENT_3 );
+	}
+
+	@Test
+	public void mustWithNotInside() {
+		assertThatQuery( index.query()
+				.where( f -> f.bool().must( f.not( f.match().field( "field2" ).matching( FIELD2_VALUE1 ) ) ) ) )
+				.hasTotalHitCount( 2 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_2, DOCUMENT_3 );
+	}
+
+	@Test
+	public void manyNestedNot() {
+		SearchPredicateFactory f = index.createScope().predicate();
+		BooleanPredicateClausesStep<?> step = f.bool().must( f.not( f.not( f.not( f.not( f.not( f.not( f.not( f.match().field( "field2" ).matching( FIELD2_VALUE1 ) ) ) ) ) ) ) ) );
+
+		// as query strings are backend specific let's check that generated string is among expected.
+		// in this case there's an odd number of nested `not` hence we should end up having mustNot/- query
+		SearchQueryOptionsStep<?, DocumentReference, StubLoadingOptionsStep, ?, ?> query = index.query()
+				.where( ( step ).toPredicate() );
+		assertThat(
+				Arrays.asList(
+						"+(-field2:[3 TO 3] #*:*)", // Lucene query
+						"{\"query\":{\"bool\":{\"must_not\":{\"match\":{\"field2\":{\"query\":3}}}}},\"_source\":false}" // Elasticsearch query
+				)
+		).contains( query.toQuery().queryString() );
+
+		assertThatQuery( query )
+				.hasTotalHitCount( 2 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_2, DOCUMENT_3 );
+
+		// as query strings are backend specific let's check that generated string is among expected.
+		// in this case there's an even number of nested `not` hence we should end up having simple match/+ query
+		query = index.query()
+				.where( f.not( ( step ) ).toPredicate() );
+		assertThat(
+				Arrays.asList(
+						"+field2:[3 TO 3]", // Lucene query
+						"{\"query\":{\"match\":{\"field2\":{\"query\":3}}},\"_source\":false}" // Elasticsearch query
+				)
+		).contains( query.toQuery().queryString() );
+
+		assertThatQuery( query )
+				.hasTotalHitCount( 1 )
+				.hasDocRefHitsAnyOrder( index.typeName(), DOCUMENT_1 );
```

Comment:
One optimization that you didn't test is the one mentioned in the ticket description:

> optimize things such as `f.bool().must(f.not(something)).must(somethingElse)` into `f.bool().mustNot(something).must(somethingElse)`

That's what implemented by the first loop turning `must` clauses into `mustNot` clauses in the predicate builders.